pkg/lib/dialogs.tsx

```diff
@@ -82,99 +82,92 @@
+ * shows a MyDialog with the given properties, plus a special "result" property
+ * which has .resolve() and .reject() methods on it.  Calling either of those
+ * will resolve the promise returned by Dialogs.run() accordingly, closing the
+ * dialog in the process.  If the dialog is closed by a call to Dialogs.close()
+ * then Dialogs.run() throws an error.  See the example:
  *
- * Calling "Dialogs.show" will render the given component as a direct
- * child of the inner-most enclosing "WithDialogs" component.  The
- * component is of course intended to be a dialog, such as
- * Patternfly's "Modal".  There is only ever one of these; a second
- * call to "show" is considered a bug and "Dialogs.close" should be called first.
- * "Dialogs.show" returns a promise that is settled by either "Dialogs.close" or
- * "Dialogs.reject".
+ * const MyDialog = ({ title, result }) => {
+ *     return (
+ *         <Modal title={title}>
+ *             <Button onClick={() => result.resolve("yes")}>Yes</Button>
+ *             <Button onClick={() => result.resolve("no")}>No</Button>
+ *         </Modal>
+ *     );
+ * };
  *
- * - Dialogs.close([args])
+ * const AsyncDialogExample = () => {
+ *     const Dialogs = useDialogs();
  *
- * Calling "Dialogs.close([args])" will close the currently open Dialog and
- * optionally resolve the promise with the provided "args".
+ *     const clicked = async () => {
+ *         try {
+ *             const result = await Dialogs.run(MyDialog, { title: "Example" });
+ *             console.log(result);
+ *         } catch (err) {
+ *         }
+ *     };
  *
- * - Dialogs.reject(err)
-  *
- * Calling "Dialogs.reject(err)" will close the currently open Dialog
- * and reject the promise with the provided Error.
+ *     return <Button onClick={clicked}>Open dialog</Button>;
+ * };
  */
 
-import React, { useContext, useRef, useState } from "react";
+import React, { useContext, useState } from "react";
 
-export const DialogsContext = React.createContext();
-export const useDialogs = () => useContext(DialogsContext);
+export interface DialogResult<T> {
+    resolve(value: T): void;
+    reject(exc: unknown): void;
+}
 
-export const WithDialogs = ({ children }) => {
-    const is_open = useRef(false); // synchronous
-    const resolveRef = useRef(null);
-    const rejectRef = useRef(null);
-    const [dialog, setDialog] = useState(null);
+export interface Dialogs {
+    show(dialog: React.ReactNode): void;
+    close(): void;
+    run<T, P>(component: React.ComponentType<P & { result: DialogResult<T> }>, properties: P): Promise<T>;
+    isActive(): boolean;
+}
 
-    const Dialogs = {
-        show: component => {
-            if (component && is_open.current)
-                console.error("Dialogs.show() called for",
-                              JSON.stringify(component),
-                              "while a dialog is already open:",
-                              JSON.stringify(dialog));
```

Comment:
Please bring back this check. This has been the cause of flakes which took me hours to understand and debug, and we don't want future code to fall into the same trap again.

Reply:
Sure.

Reply:
See above.